### DISCUSSION TITLE
kinesis graphical layout editor with miryoku

### DISCUSSION THREAD

**Discussion by symonh** [manna-harbour/miryoku#227] [Q&A] [answered]
Hi all, I'm trying to set up an advantage360 with miryoku, and I'm wondering if there is any way to use kinesis's gui to edit the layout? I really love how easy moergo has made it to import other people's layout into their gui editor — wish kinesis would do the same for we newbs and no-coders! Thanks! 

Edit: When I direct the kinesis editor to my fork of this repo, it fails because it can't find config/info.json. I tried copying in the file from the repo I forked from kinesis, but that results in a blank layout when I point the gui to my fork of miryoku

edit: Bravekarma on discord helped me with a 'decompiled' version. I added the keycap to my Kinesis-forked repo, but the keymap editor still doesn't recognize anything (no errors — it's just blank). Here's the repo: https://github.com/symonh/Adv360-Pro-ZMK-SC

Bravekarma suggested that I ask if there's a way to get "a keymap to build on top of Miryoku, while keeping editor support somehow" 

The info.json file looks like this

```

> {
>   "layouts": {
>     "LAYOUT": {
>       "layout": [
>         { "label": "=", "row": 0, "col": 0, "x": 0, "y": 0.25, "w":1.25 },
>         { "label": "1", "row": 0, "col": 1, "x": 1.25, "y": 0.25 },
>         { "label": "2", "row": 0, "col": 2, "x": 2.25, "y": 0 },
>         { "label": "3", "row": 0, "col": 3, "x": 3.25, "y": 0 },
>         { "label": "4", "row": 0, "col": 4, "x": 4.25, "y": 0 },
>         { "label": "5", "row": 0, "col": 5, "x": 5.25, "y": 0 },
>         { "label": "mod1", "row": 0, "col": 6, "x": 6.25, "y": 0 },
>         { "label": "mod2", "row": 0, "col": 7, "x": 10.75, "y": 0 },
>         { "label": "6", "row": 0, "col": 8, "x": 11.75, "y": 0 },
>         { "label": "7", "row": 0, "col": 9, "x": 12.75, "y": 0 },
>         { "label": "8", "row": 0, "col": 10, "x": 13.75, "y": 0 },
>         { "label": "9", "row": 0, "col": 11, "x": 14.75, "y": 0 },
>         { "label": "0", "row": 0, "col": 12, "x": 15.75, "y": 0.25 },
>         { "label": "-", "row": 0, "col": 13, "x": 16.75, "y": 0.25, "w": 1.25 },
> 
>         { "label": "Tab", "row": 1, "col": 0, "x": 0, "y": 1.25, "w": 1.25 },
>         { "label": "Q", "row": 1, "col": 1, "x": 1.25, "y": 1.25 },
>         { "label": "W", "row": 1, "col": 2, "x": 2.25, "y": 1 },
>         { "label": "E", "row": 1, "col": 3, "x": 3.25, "y": 1 },
>         { "label": "R", "row": 1, "col": 4, "x": 4.25, "y": 1 },
>         { "label": "T", "row": 1, "col": 5, "x": 5.25, "y": 1 },
>         { "label": "mod3", "row": 1, "col": 6, "x": 6.25, "y": 1 },
>         { "label": "mod4", "row": 1, "col": 7, "x": 10.75, "y": 1 },
>         { "label": "Y", "row": 1, "col": 8, "x": 11.75, "y": 1 },
>         { "label": "U", "row": 1, "col": 9, "x": 12.75, "y": 1 },
>         { "label": "I", "row": 1, "col": 10, "x": 13.75, "y": 1 },
>         { "label": "O", "row": 1, "col": 11, "x": 14.75, "y": 1 },
>         { "label": "P", "row": 1, "col": 12, "x": 15.75, "y": 1.25 },
>         { "label": "\\", "row": 1, "col": 13, "x": 16.75, "y": 1.25, "w": 1.25 },
> 
> 
>         { "label": "Caps", "row": 2, "col": 0, "x": 0, "y": 2.25, "w":1.25 },
>         { "label": "A", "row": 2, "col": 1, "x": 1.25, "y": 2.25 },
>         { "label": "S", "row": 2, "col": 2, "x": 2.25, "y": 2 },
>         { "label": "D", "row": 2, "col": 3, "x": 3.25, "y": 2 },
>         { "label": "F", "row": 2, "col": 4, "x": 4.25, "y": 2 },
>         { "label": "G", "row": 2, "col": 5, "x": 5.25, "y": 2 },
>         { "label": "mod5", "row": 2, "col": 6, "x": 6.25, "y": 2},
>         { "label": "LCtrl", "row": 2, "col": 7, "x": 6.75, "y": 4, "r":15, "rx":5.25 ,"ry":4},
>         { "label": "LAlt", "row": 2, "col": 8, "x": 7.75, "y": 4, "r":15, "rx":5.25 ,"ry":4},
>         { "label": "LGui", "row": 2, "col": 11, "x": 9.25, "y": 4, "r":-15, "rx":12.75 ,"ry":4},
>         { "label": "RCtrl", "row": 2, "col": 12, "x": 10.25, "y": 4, "r":-15, "rx":12.75 ,"ry":4},
>         { "label": "mod6", "row": 2, "col": 13, "x": 10.75, "y": 2},
>         { "label": "H", "row": 2, "col": 14, "x": 11.75, "y": 2 },
>         { "label": "J", "row": 2, "col": 15, "x": 12.75, "y": 2 },
>         { "label": "K", "row": 2, "col": 16, "x": 13.75, "y": 2 },
>         { "label": "L", "row": 2, "col": 17, "x": 14.75, "y": 2 },
>         { "label": ";", "row": 2, "col": 18, "x": 15.75, "y": 2.25 },
>         { "label": "'", "row": 2, "col": 19, "x": 16.75, "y": 2.25, "w":1.25 },
> 
>         { "label": "LShift", "row": 3, "col": 0, "x": 0, "y": 3.25, "w":1.25 },
>         { "label": "Z", "row": 3, "col": 1, "x": 1.25, "y": 3.25 },
>         { "label": "X", "row": 3, "col": 2, "x": 2.25, "y": 3 },
>         { "label": "C", "row": 3, "col": 3, "x": 3.25, "y": 3 },
>         { "label": "V", "row": 3, "col": 4, "x": 4.25, "y": 3 },
>         { "label": "B", "row": 3, "col": 5, "x": 5.25, "y": 3 },
>         { "label": "1U1", "row": 3, "col": 6, "x": 5.75, "y": 5, "r":15, "rx":5.25 ,"ry":4, "h":0},
>         { "label": "1U2", "row": 3, "col": 7, "x": 6.75, "y": 5, "r":15, "rx":5.25 ,"ry":4, "h":0},
>         { "label": "Home", "row": 3, "col": 8, "x": 7.75, "y": 5, "r":15, "rx":5.25 ,"ry":4},
>         { "label": "Ped3", "row": 3, "col": 9, "x": 20, "y": 5},
>         { "label": "Ped4", "row": 3, "col": 10, "x": 21, "y": 5},
>         { "label": "PgUp", "row": 3, "col": 11, "x": 9.25, "y": 5, "r":-15, "rx":12.75 ,"ry":4},
>         { "label": "1U3", "row": 3, "col": 12, "x": 10.25, "y": 5, "r":-15, "rx":12.75 ,"ry":4, "h":0},
>         { "label": "1U4", "row": 3, "col": 13, "x": 11.25, "y": 5, "r":-15, "rx":12.75 ,"ry":4, "h":0},
>         { "label": "N", "row": 3, "col": 14, "x": 11.75, "y": 3 },
>         { "label": "M", "row": 3, "col": 15, "x": 12.75, "y": 3 },
>         { "label": ",", "row": 3, "col": 16, "x": 13.75, "y": 3 },
>         { "label": ".", "row": 3, "col": 17, "x": 14.75, "y": 3 },
>         { "label": "/", "row": 3, "col": 18, "x": 15.75, "y": 3.25 },
>         { "label": "RShift", "row": 3, "col": 19, "x": 16.75, "y": 3.25, "w":1.25 },
> 
>         { "label": "mod7", "row": 4, "col": 0, "x": 0, "y": 4.25, "w":1.25},
>         { "label": "`", "row": 4, "col": 1, "x": 1.25, "y": 4.25 },
>         { "label": "\\", "row": 4, "col": 2, "x": 2.25, "y": 4 },
>         { "label": "V", "row": 4, "col": 3, "x": 3.25, "y": 4 },
>         { "label": "B", "row": 4, "col": 4, "x": 4.25, "y": 4 },
>         { "label": "Bksp", "row": 4, "col": 5, "x": 5.75, "y": 5, "r":15, "rx":5.25 ,"ry":4, "h":2},
>         { "label": "Del", "row": 4, "col": 6, "x": 6.75, "y": 5, "r":15, "rx":5.25 ,"ry":4, "h":2},
>         { "label": "End", "row": 4, "col": 7, "x": 7.75, "y": 6, "r":15, "rx":5.25 ,"ry":4},
>         { "label": "PgDn", "row": 4, "col": 8, "x": 9.25, "y": 6, "r":-15, "rx":12.75 ,"ry":4},
>         { "label": "Enter", "row": 4, "col": 9, "x": 10.25, "y": 5, "r":-15, "rx":12.75 ,"ry":4, "h":2},
>         { "label": "Space", "row": 4, "col": 10, "x": 11.25, "y": 5, "r":-15, "rx":12.75 ,"ry":4, "h":2},
>         { "label": "Up", "row": 4, "col": 11, "x": 12.75, "y": 4 },
>         { "label": "Down", "row": 4, "col": 12, "x": 13.75, "y": 4 },
>         { "label": ",", "row": 4, "col": 13, "x": 14.75, "y": 4 },
>         { "label": ".", "row": 4, "col": 14, "x": 15.75, "y": 4.25 },
>         { "label": "mod8", "row": 4, "col": 15, "x": 16.75, "y": 4.25, "w":1.25 }
>       ]
>     }
>   }
> }

```

**Comment by manna-harbour** [comment]
You can build Miryoku ZMK for `adv360pro`, and customise layers or map unused keys using https://github.com/manna-harbour/miryoku/discussions/85.

If the GUI editor can import ZMK keymap files then it should be possible to import the Miryoku ZMK keymap by first expanding all of the Miryoku includes and macros using `cpp`.

There's no support for generating the editor's json file directly, although it could be done starting with Miryoku Babel as with other Miryoku implementations.

**Reply by symonh** [reply]
Thank you for this information and for the quick reply. Unfortunately, I'm not knowledgeable enough to understand how to do that stuff using the instructions at #85, and unfortunately I don't have time to learn how to do this. I really wish Kinesis had done something like what Moergo did with the Glove80 — they let you share, import, and edit layouts on their website and it works very nicely. Generally disappointed by Kinesis's support for this wildly expensive keyboard and would gladly switch to the Glove80 if it didn't use low-profile keys :/ Thanks again for your help — maybe I'll figure this stuff out when I have more time.

**Reply by manna-harbour** [reply]
You can build with any combination of alternative layout options without having to make any edits. See the [Miryoku ZMK Quickstart Guide](https://github.com/manna-harbour/miryoku_zmk/tree/master/docs/quickstart).

To customise layers or map unused keys you will have to edit a file as described above. There's also a step-by-step example of customising a layer at https://github.com/manna-harbour/miryoku/discussions/224#discussioncomment-6323631.

Also, support for `adv360` (requires Kinesis fork) has been added to Miryoku ZMK, in addition to `adv360pro` (builds with ZMK main). Here's a sample workflow build for both variants: https://github.com/manna-harbour/miryoku_zmk/actions/runs/5531180161.

**Comment by manna-harbour** [comment, marked as answer]
- select alternative layout options via `custom_config.h`
- generate a single file keymap with `pcpp --passthru-unfound-includes adv360.keymap` in the Miryoku ZMK `config` directory
- import the keymap file into the editor
- optionally add the editor generated keymap to your Miryoku ZMK fork and build for `adv360` and it will build using the Kinesis fork automatically

Here is a sample single file keymap: [adv360.keymap.txt](https://github.com/manna-harbour/miryoku/files/12028296/adv360.keymap.txt).

**Reply by manna-harbour** [reply]
More info from discord comment.

>Install `pcpp` (`pip install pcpp` or similar), edit `custom_config.h` according to the Miryoku ZMK docs, `cd` to the `config` directory in your clone, run the command line mentioned previously, and use the output of that with the editor. This has to be done locally at the moment but I can add it to the workflow later.
Then use the editor generated keymap to build as usual, either locally or with a workflow, via Miryoku ZMK or a Kinesis repo.
